makelib/imagelight.mk

```diff
@@ -0,0 +1,214 @@
+CACHE_REGISTRY := cache
+
+# prune images that are at least this many hours old
+PRUNE_HOURS ?= 48
+
+# prune keeps at least this many images regardless of how old they are
+PRUNE_KEEP ?= 24
+
+# don't actually prune just show what prune would do.
+PRUNE_DRYRUN ?= 0
+
+# the cached image format
+CACHE_DATE_FORMAT := "%Y-%m-%d.%H%M%S"
+CACHE_PRUNE_DATE := $(shell export TZ="UTC+$(PRUNE_HOURS)"; date +"$(CACHE_DATE_FORMAT)")
+CACHE_TAG := $(shell date -u +"$(CACHE_DATE_FORMAT)")
+
+REGISTRY_ORGS ?= docker.io
+IMAGE_ARCHS := $(subst linux_,,$(filter linux_%,$(PLATFORMS)))
+IMAGE_PLATFORMS := $(subst _,/,$(subst $(SPACE),$(COMMA),$(filter linux_%,$(PLATFORMS))))
+IMAGE_PLATFORMS_LIST := $(subst _,/,$(filter linux_%,$(PLATFORMS)))
+IMAGE_PLATFORM := $(subst _,/,$(PLATFORM))
+
+# if set to 1 docker image caching will not be used.
+CACHEBUST ?= 0
+ifeq ($(CACHEBUST),1)
+BUILD_ARGS += --no-cache
+endif
+
+ifeq ($(HOSTOS),Linux)
+SELF_CID := $(shell cat /proc/self/cgroup | grep docker | grep -o -E '[0-9a-f]{64}' | head -n 1)
+endif
+
+# =====================================================================================
+# Image Targets
+
+do.img.clean:
+	@for i in $(CLEAN_IMAGES); do \
+		if [ -n "$$(docker images -q $$i)" ]; then \
+			for c in $$(docker ps -a -q --no-trunc --filter=ancestor=$$i); do \
+				if [ "$$c" != "$(SELF_CID)" ]; then \
+					echo stopping and removing container $${c} referencing image $$i; \
+					docker stop $${c}; \
+					docker rm $${c}; \
+				fi; \
+			done; \
+			echo cleaning image $$i; \
+			docker rmi $$i > /dev/null 2>&1 || true; \
+		fi; \
+	done
+
+# this will clean everything for this build
+img.clean:
+	@$(INFO) cleaning images for $(BUILD_REGISTRY)
+	@$(MAKE) do.img.clean CLEAN_IMAGES="$(shell docker images | grep -E '^$(BUILD_REGISTRY)/' | awk '{print $$1":"$$2}')"
+	@$(OK) cleaning images for $(BUILD_REGISTRY)
+
+img.done:
+	@rm -fr $(IMAGE_TEMP_DIR)
+
+img.cache:
+	@for i in $(CACHE_IMAGES); do \
+		IMGID=$$(docker images -q $$i); \
+		if [ -n "$$IMGID" ]; then \
+			echo === caching image $$i; \
+			CACHE_IMAGE=$(CACHE_REGISTRY)/$${i#*/}; \
+			docker tag $$i $${CACHE_IMAGE}:$(CACHE_TAG); \
+			for r in $$(docker images --format "{{.ID}}#{{.Repository}}:{{.Tag}}" | grep $$IMGID | grep $(CACHE_REGISTRY)/ | grep -v $${CACHE_IMAGE}:$(CACHE_TAG)); do \
+				docker rmi $${r#*#} > /dev/null 2>&1 || true; \
+			done; \
+		fi; \
+	done
+
+# prune removes old cached images
+img.prune:
+	@$(INFO) pruning images older than $(PRUNE_HOURS) keeping a minimum of $(PRUNE_KEEP) images
+	@EXPIRED=$$(docker images --format "{{.Tag}}#{{.Repository}}:{{.Tag}}" \
+		| grep -E '$(CACHE_REGISTRY)/' \
+		| sort -r \
+		| awk -v i=0 -v cd="$(CACHE_PRUNE_DATE)" -F  "#" '{if ($$1 <= cd && i >= $(PRUNE_KEEP)) print $$2; i++ }') &&\
+	for i in $$EXPIRED; do \
+		echo removing expired cache image $$i; \
+		[ $(PRUNE_DRYRUN) = 1 ] || docker rmi $$i > /dev/null 2>&1 || true; \
+	done
```

Comment:
I noticed that the cross image uses this cache functionality, so some of this pruning logic would skip that now -- however, part of the point of this change is that we are no longer building the cross image, so I am not too concerned about removing all of the existing cache stuff